### PR TITLE
Add AWS ARN for the specific offending Resource into the Description of the ASFF parser

### PR DESCRIPTION
**Pull request by testaccount90009**
## :warning: Note on feature completeness :warning:

We are narrowing the scope of acceptable enhancements to DefectDojo in preparation for v3. Learn more here:
https://github.com/DefectDojo/django-DefectDojo/blob/master/readme-docs/CONTRIBUTING.md

**Description**

This implements the AWS ARN and Resource specifically associated with the Finding and the Security Control description.  Without this, it is **extremely** difficult at a large scale to decipher the specific resource within the failing control from Security Hub.

**I have already tested this in my Dev DefectDojo stack** and imported a test file.  It works completely fine and adds both the AWS ARN and the SecurityControl description to the Description field in DefectDojo with the Finding, exactly as intended.

I don't know if this is a BugFix, Enhancement, or other label. However, in a perfect world (prior to this pull request) the DefectDojo system would indeed capture the offending AWS ARN associated with the failing Security Control description.  The description alone is not sufficient in triaging vulnerabilities and providing information on **what** the offending resource actually is.

**Test results**

Ideally you extend the test suite in `tests/` and `dojo/unittests` to cover the changed in this PR.
Alternatively, describe what you have and haven't tested.

I have not changed or added to any tests, since I am only grabbing an additional field (Resources:0/Id) and adding it to the description.  If it does not exist, it simply reverts to the original 'description' of the Security Control.

**Documentation**

Please update any documentation when needed in the [documentation folder](https://github.com/DefectDojo/django-DefectDojo/tree/dev/docs))

**Checklist**

This checklist is for your information.

- [X] Make sure to rebase your PR against the very latest `dev`.
- [X] Features/Changes should be submitted against the `dev`.
- [ ] Bugfixes should be submitted against the `bugfix` branch.
- [X] Give a meaningful name to your PR, as it may end up being used in the release notes.
- [ ] Your code is flake8 compliant.
- [X] Your code is python 3.11 compliant.
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the docs at https://github.com/DefectDojo/django-DefectDojo/tree/dev/docs as part of this PR.
- [ ] Model changes must include the necessary migrations in the dojo/db_migrations folder.
- [ ] Add applicable tests to the unit tests.
- [ ] Add the proper label to categorize your PR.

**Extra information**

Please clear everything below when submitting your pull request, it's here purely for your information.

Moderators: Labels currently accepted for PRs:
- Import Scans (for new scanners/importers)
- enhancement
- performance
- feature
- bugfix
- maintenance (a.k.a chores)
- dependencies
- New Migration (when the PR introduces a DB migration)
- settings_changes (when the PR introduces changes or new settings in settings.dist.py)

# Contributors: Git Tips
## Rebase on dev branch
If the dev branch has changed since you started working on it, please rebase your work after the current dev.

On your working branch `mybranch`:
```
git rebase dev mybranch
```
In case of conflict:
```
 git mergetool
 git rebase --continue
 ```

When everything's fine on your local branch, force push to your `myOrigin` remote:
```
git push myOrigin --force-with-lease
```

To cancel everything:
```
git rebase --abort
```

## Squashing commits
```
git rebase -i origin/dev
```
- Replace `pick` by `fixup` on the commits you want squashed out
- Replace `pick` by `reword` on the first commit if you want to change the commit message
- Save the file and quit your editor

Force push to your `myOrigin` remote:
```
git push myOrigin --force-with-lease
```

### CODE REVIEW

**Comment by testaccount90009** [comment]
thumbnail

**Comment by testaccount90009** [comment]
image

File: dojo/tools/asff/parser.py [hunk 1]
```diff
@@ -41,9 +41,29 @@ def get_findings(self, file, test):
...
+            # control and has no information about the offending resource.
+            #
+            # Retrieve the AWS ARN / Resource Id
+            resource_arn = item.get("ProductFields").get("Resources:0/Id")
+
+            # Define the control_description
+            control_description = item.get("Description")
+
+            if resource_arn:
+                full_description = f"{resource_arn}\n\n{control_description}"
```
Comment (cneill) [inline dojo/tools/asff/parser.py:60]: Could you add some kind of indication of what this is? E.g. `**AWS resource ARN:** {resource_arn}`

File: dojo/tools/asff/parser.py [hunk 2]
```diff
@@ -41,9 +41,29 @@ def get_findings(self, file, test):
...
+            #
+            # This is needed because every Finding in AWS from Security Hub has an
+            # associated ResourceId that contains the full AWS ARN and without it,
+            # it is much more difficult to track down the specific resource.
+            #
+            # This is different from the Finding Id - as that is from the Security Hub
+            # control and has no information about the offending resource.
+            #
+            # Retrieve the AWS ARN / Resource Id
+            resource_arn = item.get("ProductFields").get("Resources:0/Id")
```
Comment (cneill) [inline dojo/tools/asff/parser.py:54]: `ProductFields` is an [optional ASFF field](https://docs.aws.amazon.com/securityhub/latest/userguide/asff-top-level-attributes.html#asff-productfields). `Resources`, on the other hand, is a [required field](https://docs.aws.amazon.com/securityhub/latest/userguide/asff-required-attributes.html#Resources) and might make more sense to reference here. You can also get more than one ARN more easily, if applicable to a given finding.

File: dojo/tools/asff/parser.py (hunk 1 above)
Comment (testaccount90009) [reply dojo/tools/asff/parser.py]: Thank you for the suggestion - I went ahead and added that!

File: dojo/tools/asff/parser.py (hunk 2 above)
Comment (testaccount90009) [reply dojo/tools/asff/parser.py]: Good catch - this is the exact same value, done in a much better way.  Thank you!

**Comment by testaccount90009** [comment]
It seems there are errors in the unit tests for the self assertion with the previous description and it no longer matches with this change.  Any advice on how to edit the unit test file to get the self assertion error to pick up the correct ResourceId?  I see them being displayed as != the previous Description, but unsure of where to go from here.

Advice and input would be greatly appreciated.

File: dojo/tools/asff/parser.py [hunk 3]
```diff
@@ -41,9 +41,35 @@ def get_findings(self, file, test):
...
+            # This is different from the Finding Id - as that is from the Security Hub
+            # control and has no information about the offending resource.
+            #
+            # Retrieve the AWS ARN / Resource Id
+            resource_arns = []
+            resource_ids = item.get("Resources", [])
+            for resource in resource_ids:
+                resource_id = resource.get("Id")
+                if resource_id:
+                    resource_arns.append(resource_id)
```
Comment (cneill) [inline dojo/tools/asff/parser.py:59]: To fix the unit tests, you'll want to update the line [here](https://github.com/DefectDojo/django-DefectDojo/blob/b25c66676d4f0a88dff66911960ebb78bbf26715/unittests/tools/test_asff_parser.py#L21) to use the new description format. You could refactor the parser changes a bit, using a helper method to retrieve the ARNs from a given `item` and leveraging that in the unit test along with the format string that actually populates it in the description so you don't have to copy the whole implementation over. Roughly something like:

```python
def get_item_resource_arns(self, item):
    ... your current code for walking the item's resources to collect all the ARNs on line 54-59 ...
    return resource_arns
```

And then reference that here:

```suggestion
            resource_arns = self.get_item_resource_arns(item)
```

Then do the same in the unit test, collecting the ARNs and referencing them in a new description format string to match the parser's formatting. You'll need to pass `parser` into `common_check_finding` in the calls from the [`test_asff_one_vuln`](https://github.com/DefectDojo/django-DefectDojo/blob/b25c66676d4f0a88dff66911960ebb78bbf26715/unittests/tools/test_asff_parser.py#L43), [`test_asff_many_vulns`](https://github.com/DefectDojo/django-DefectDojo/blob/b25c66676d4f0a88dff66911960ebb78bbf26715/unittests/tools/test_asff_parser.py#L52), and [`test_asff_guardduty`](https://github.com/DefectDojo/django-DefectDojo/blob/b25c66676d4f0a88dff66911960ebb78bbf26715/unittests/tools/test_asff_parser.py#L61) methods so you can reference `parser.get_item_resource_arns` from `common_check_finding`.

Comment (testaccount90009) [reply dojo/tools/asff/parser.py]: I believe I have refactored the parser as you have suggested (thank you again!).  I also passed in the parser = AsffParser() on line 20 in the common_check_finding.

I'm now in the process of going through those load_sample_json files and collecting the ARNs so I can reference them in the new description format string.

Would this change take place on L22 and editing:
self.assertEqual(finding.description, data[index]["Description"])

or do I need to capture each respective load_sample_json ARN + description and within those methods pass a self.assertEqual value for the finding.description with the new value format of:
"**AWS resource ARN:** {resource_arn_strings}\n\n{control_description}"

**Comment by testaccount90009** [comment]
I spent some time trying to work in the suggestions for the unit tests @cneill , but unfortunately I was not able to figure out how to properly do that.

**Comment by testaccount90009** [comment]
@cneill I guess I ended up getting it.  Broken clock is right twice a day!  

I appreciate your guidance and the opportunity to learn with this.  This is my first ever unit test and parser change, so I have learned a lot from this.

Many thanks!

**Review by mtesauro — APPROVED** [review]
Approved

**Review by cneill — APPROVED** [review]
Looks good! Happy to help 🙂

**Review by Maffooch — APPROVED** [review]

**Comment by testaccount90009** [comment]
So upon testing this, it seems that items with multiple ARNs joined by a ',' in the description will break traditional .csv parsers (at least ours did).

I am considering replacing the ',' join with either a '+' or '&' instead.

Thoughts or feedback?

**Pull request merged**